packages/vite-plugin-svelte/src/utils/vite-plugin-svelte-stats.ts

```diff
@@ -0,0 +1,206 @@
+					} else {
+						return cell.padStart(columnWidths[i], ' ');
+					}
+				})
+				.join('\t')
+		)
+		.join('\n');
+	return table;
+}
+
+export class VitePluginSvelteStats {
+	// package directory -> package name
+	private _packages: { path: string; name: string }[] = [];
+	private _collections: StatCollection[] = [];
+	startCollection(name: string, opts?: Partial<CollectionOptions>) {
+		const options = {
+			...defaultCollectionOptions,
+			...opts
+		};
+		const stats: Stat[] = [];
+		const collectionStart = performance.now();
+		const _this = this;
+		let hasLoggedProgress = false;
+		const collection: StatCollection = {
+			name,
+			options,
+			stats,
+			collectionStart,
+			finished: false,
+			start(file) {
+				if (collection.finished) {
+					throw new Error('called after finish() has been used');
+				}
+				const start = performance.now();
+				const stat: Stat = { file, start, end: start };
+				return () => {
+					const now = performance.now();
+					stat.end = now;
+					stats.push(stat);
+					if (!hasLoggedProgress && options.logInProgress(collection, now)) {
+						hasLoggedProgress = true;
+						log.info(`${name} in progress ...`);
+					}
+				};
+			},
+			async finish() {
+				await _this._finish(collection);
+			}
+		};
+		_this._collections.push(collection);
+		return collection;
+	}
+
+	public async finishAll() {
+		await Promise.all(this._collections.map((c) => c.finish()));
+	}
+
+	private async _finish(collection: StatCollection) {
+		collection.finished = true;
+		const now = performance.now();
+		collection.duration = now - collection.collectionStart;
+		const logResult = collection.options.logResult(collection);
+		if (logResult) {
+			await this._aggregateStatsResult(collection);
+			log.info(`${collection.name} done.`, formatPackageStats(collection.packageStats!));
+		}
+		// cut some ties to free it for garbage collection
+		const index = this._collections.indexOf(collection);
+		this._collections.splice(index, 1);
+		collection.stats.length = 0;
+		collection.stats = [];
+		if (collection.packageStats) {
+			collection.packageStats.length = 0;
+			collection.packageStats = [];
+		}
+		collection.start = () => () => {};
+		collection.finish = () => {};
+	}
+
+	private async _aggregateStatsResult(collection: StatCollection) {
+		const stats = collection.stats;
+		for (const stat of stats) {
+			let pkg = this._packages.find((p) => stat.file.startsWith(p.path));
+			if (!pkg) {
+				// check for package.json first
+				let pkgPath = await findClosestPkgJsonPath(stat.file);
+				if (pkgPath) {
+					let path = pkgPath?.replace(/package.json$/, '');
+					let name = JSON.parse(readFileSync(pkgPath, 'utf-8')).name;
+					if (!name) {
+						// some packages have nameless nested package.json
+						pkgPath = await findClosestPkgJsonPath(path);
+						if (pkgPath) {
+							path = pkgPath?.replace(/package.json$/, '');
+							name = JSON.parse(readFileSync(pkgPath, 'utf-8')).name;
+						}
+					}
```

Comment:
This is.. interesting 😄 I guess it would be nice if vitefu exported a more flexible API for this, but I don't usually mind code like this for now.

Reply:
yeah, not proud of that one, but it works :neutral_face: